File: src/main/java/me/tomassetti/examples/MethodUsage.java
package me.tomassetti.examples;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.MethodCallExpr;
import me.tomassetti.support.SpecificNodeIterator;
import me.tomassetti.symbolsolver.javaparsermodel.JavaParserFacade;
import me.tomassetti.symbolsolver.model.declarations.MethodDeclaration;
import me.tomassetti.symbolsolver.model.resolution.SymbolReference;
import me.tomassetti.symbolsolver.model.resolution.TypeSolver;
import me.tomassetti.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
import me.tomassetti.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
import me.tomassetti.symbolsolver.resolution.typesolvers.JreTypeSolver;
import java.io.File;
import me.tomassetti.support.DirExplorer;

import java.io.FileNotFoundException;
import java.util.LinkedList;
import java.util.List;

public class MethodUsage {
    private static int solved = 0;
    private static int unsolved = 0;
    private static int errors = 0;

    public static List<File> findJavaFiles(File projectDir) {
        LinkedList<File> files = new LinkedList<File>();
        new DirExplorer((level, path, file) -> path.endsWith(".java"), (level, path, file) -> {
            //System.out.println(path);
            //System.out.println(Strings.repeat("=", path.length()));
            files.add(file);
        }).explore(projectDir);
        return files;
    }

    public static List<MethodCallExpr> descendantsOfType(CompilationUnit cu) {
        LinkedList<MethodCallExpr> descendants = new LinkedList<>();
        new SpecificNodeIterator<MethodCallExpr>(
                new SpecificNodeIterator.NodeHandler<MethodCallExpr>() {
                    @Override
                    public boolean handle(MethodCallExpr node) {
                        descendants.add(node);
                        return true;
                    }
                },
                MethodCallExpr.class
        ).explore(cu);

        return descendants;
    }

    public static void processJavaFile(File file, JavaParserFacade javaParserFacade) throws FileNotFoundException {
        System.out.println(file);
        descendantsOfType(JavaParser.parse(file)).forEach(
                it->{
                    System.out.print(" * L"+it.getBegin().line+" "+ it+ " ");
                    try {
                        SymbolReference<MethodDeclaration> methodRef = javaParserFacade.solve(it);
                        if (methodRef.isSolved()) {
                            solved++;
                            MethodDeclaration methodDecl = methodRef.getCorrespondingDeclaration();
                            System.out.println("  -> "+methodDecl.getQualifiedSignature());
                        } else {
                            unsolved++;
                            System.out.println(" ???");
                        }
                    } catch (Exception e) {
                        System.out.println(" ERR "+e.getMessage());
                        errors++;
                    } catch (Throwable t) {
                        t.printStackTrace();;
                    }
                }
        );
    }

    public static TypeSolver typeSolver() {
        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver();
        combinedTypeSolver.add(new JreTypeSolver());
        combinedTypeSolver.add(
                new JavaParserTypeSolver(new File("src/main/resources/javaparser-core"))
        );
        combinedTypeSolver.add(
                new JavaParserTypeSolver(new File("src/main/resources/javaparser-generated-sources"))
        );
        return combinedTypeSolver;
    }

    public static void main(String[] args) {
        File projectDir = new File("src/main/resources/javaparser-core");
        List<File> javaFiles = findJavaFiles(projectDir);
        JavaParserFacade javaParserFacade = JavaParserFacade.get(typeSolver());
        javaFiles.forEach(it -> {
            try {
                processJavaFile(it, javaParserFacade);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        });
        System.out.println("solved "+solved+" unsolved "+unsolved+" errors "+errors);
    }
}

File: src/main/kotlin/MethodUsage.kt

import com.github.javaparser.JavaParser
import com.github.javaparser.ast.expr.MethodCallExpr
import me.tomassetti.symbolsolver.javaparsermodel.JavaParserFacade
import me.tomassetti.symbolsolver.model.resolution.TypeSolver
import me.tomassetti.symbolsolver.resolution.typesolvers.CombinedTypeSolver
import me.tomassetti.symbolsolver.resolution.typesolvers.JavaParserTypeSolver
import me.tomassetti.symbolsolver.resolution.typesolvers.JreTypeSolver
import java.io.File

fun typeSolver() : TypeSolver {
    val combinedTypeSolver = CombinedTypeSolver()
    combinedTypeSolver.add(JreTypeSolver())
    combinedTypeSolver.add(JavaParserTypeSolver(File("src/main/resources/javaparser-core")))
    combinedTypeSolver.add(JavaParserTypeSolver(File("src/main/resources/javaparser-generated-sources")))
    return combinedTypeSolver
}

class DirExplorer(private val filter: DirExplorer.Filter, private val fileHandler: DirExplorer.FileHandler) {
    interface FileHandler {
        fun handle(level: Int, path: String, file: File)
    }

    interface Filter {
        fun interested(level: Int, path: String, file: File): Boolean
    }

    fun explore(root: File) {
        explore(0, "", root)
    }

    private fun explore(level: Int, path: String, file: File) {
        if (file.isDirectory) {
            for (child in file.listFiles()) {
                explore(level + 1, path + "/" + child.name, child)
            }
        } else {
            if (filter.interested(level, path, file)) {
                fileHandler.handle(level, path, file)
            }
        }
    }

}

var solved = 0
var unsolved = 0
var errors = 0

fun processJavaFile(file: File, javaParserFacade: JavaParserFacade) {
    println(file)
    JavaParser.parse(file).descendantsOfType(MethodCallExpr::class.java).forEach {
        print(" * L${it.begin.line} $it ")
        try {
            val methodRef = javaParserFacade.solve(it)
            if (methodRef.isSolved) {
                solved++
                val methodDecl = methodRef.correspondingDeclaration
                println("  -> ${methodDecl.qualifiedSignature}")
            } else {
                unsolved++
                println(" ???")
            }
        } catch (e: Exception) {
            println(" ERR ${e.message}")
            errors++
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }
}

fun main(args:Array<String>) {
    val javaFiles = findJavaFiles(File("src/main/resources/javaparser-core"))
    val javaParserFacade = JavaParserFacade.get(typeSolver())
    javaFiles.forEach { processJavaFile(it, javaParserFacade) }
    println("solved $solved unsolved $unsolved errors $errors")
}